File: backend/app/schemas/meeting.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class ParticipantBase(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    email: str | None = None


class ParticipantCreate(ParticipantBase):
    pass


class ParticipantRead(ParticipantBase):
    model_config = ConfigDict(from_attributes=True)
    id: int


class TagRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str
    color: str


class TranscriptSegmentBase(BaseModel):
    speaker: str
    start_ms: int = Field(ge=0)
    end_ms: int = Field(ge=0)
    text: str
    segment_order: int = 0


class TranscriptSegmentCreate(TranscriptSegmentBase):
    pass


class TranscriptSegmentRead(TranscriptSegmentBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    meeting_id: int


class ActionItemBase(BaseModel):
    text: str
    assignee: str | None = None
    due_date: date | None = None
    completed: bool = False


class ActionItemCreate(ActionItemBase):
    pass


class ActionItemUpdate(BaseModel):
    text: str | None = None
    assignee: str | None = None
    due_date: date | None = None
    completed: bool | None = None


class ActionItemRead(ActionItemBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    meeting_id: int


class TopicBase(BaseModel):
    title: str
    description: str | None = None
    start_ms: int = 0


class TopicCreate(TopicBase):
    pass


class TopicRead(TopicBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    meeting_id: int


class MeetingBase(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    meeting_date: datetime
    duration_seconds: int = Field(ge=0, default=0)
    media_url: str | None = None
    summary: str | None = None


class MeetingCreate(MeetingBase):
    participant_names: list[str] = []
    tag_names: list[str] = []
    transcript_segments: list[TranscriptSegmentCreate] = []
    action_items: list[ActionItemCreate] = []
    topics: list[TopicCreate] = []


class MeetingUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    meeting_date: datetime | None = None
    duration_seconds: int | None = None
    media_url: str | None = None
    summary: str | None = None
    participant_names: list[str] | None = None
    tag_names: list[str] | None = None


class MeetingListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    title: str
    meeting_date: datetime
    duration_seconds: int
    summary: str | None
    participants: list[ParticipantRead]
    tags: list[TagRead]


class MeetingDetail(MeetingBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    created_at: datetime
    updated_at: datetime
    participants: list[ParticipantRead]
    tags: list[TagRead]
    transcript_segments: list[TranscriptSegmentRead]
    action_items: list[ActionItemRead]
    topics: list[TopicRead]


class GlobalSearchResult(BaseModel):
    meeting_id: int
    meeting_title: str
    match_type: str
    snippet: str
    segment_id: int | None = None
    start_ms: int | None = None

File: frontend/src/types/meeting.ts
export interface Participant {
  id: number;
  name: string;
  email?: string | null;
}

export interface Tag {
  id: number;
  name: string;
  color: string;
}

export interface TranscriptSegment {
  id: number;
  meeting_id: number;
  speaker: string;
  start_ms: number;
  end_ms: number;
  text: string;
  segment_order: number;
}

export interface ActionItem {
  id: number;
  meeting_id: number;
  text: string;
  assignee?: string | null;
  due_date?: string | null;
  completed: boolean;
}

export interface Topic {
  id: number;
  meeting_id: number;
  title: string;
  description?: string | null;
  start_ms: number;
}

export interface MeetingListItem {
  id: number;
  title: string;
  meeting_date: string;
  duration_seconds: number;
  summary?: string | null;
  participants: Participant[];
  tags: Tag[];
}

export interface MeetingDetail extends MeetingListItem {
  description?: string | null;
  media_url?: string | null;
  created_at: string;
  updated_at: string;
  transcript_segments: TranscriptSegment[];
  action_items: ActionItem[];
  topics: Topic[];
}

export interface GlobalSearchResult {
  meeting_id: number;
  meeting_title: string;
  match_type: string;
  snippet: string;
  segment_id?: number | null;
  start_ms?: number | null;
}

export interface AskQuestionResponse {
  answer: string;
  source_segments: TranscriptSegment[];
  used_llm: boolean;
}

export interface MeetingCreatePayload {
  title: string;
  description?: string;
  meeting_date: string;
  duration_seconds?: number;
  media_url?: string;
  summary?: string;
  participant_names?: string[];
  tag_names?: string[];
  transcript_segments?: Omit<TranscriptSegment, "id" | "meeting_id">[];
  action_items?: Omit<ActionItem, "id" | "meeting_id">[];
  topics?: Omit<Topic, "id" | "meeting_id">[];
}
